allow customizing the inline repr of a duck array

keewis opened the pull request:
This calls the duck array's `_repr_short_` (does not have to be that name, might also be `_repr_inline_` or something else) if it exists, which can make e.g. `pint.Quantity`'s inline repr a lot more useful. We might also be able to use that to get rid of our `inline_dask_repr` and `inline_sparse_repr` functions by pushing that upstream (or by monkeypatching).

 - [x] possibly closes #2773
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
 - [ ] New functions/methods are listed in `api.rst`

max-sixty commented:
Looks great to me.

I'm not sure the naming convention: is a single `_` dunder standard? Or does that look like a private method?

jthielen commented:
I think the `_repr_*_` is the standard from IPython/Jupyter (https://ipython.readthedocs.io/en/stable/config/integrating.html), so `_repr_short_` or `_repr_inline_` is just following along with that.

keewis commented:
> I think `_repr_*_` is the standard from IPython/Jupyter

exactly, that's where the name came from.

Also, if consistency with the functions in `xarray.core.formatting` is important it might be better to choose `_repr_inline_`: the `short_*_repr` functions simply shorten the normal `repr`.

keewis commented:
if I remember correctly, this was accepted in this week's community meeting. Should this be merged, then?

max-sixty commented:
LGTM! A test wouldn't go amiss but fine to keep moving given the size of the change imo

keewis commented:
tests are a good point, and I guess this should also be documented somewhere. Any ideas where that could be? Internals?

dcherian commented:
"internals" sounds good to me. It could also go in the future NEP-18 wrapping section.

keewis commented:
I put it into internals for now, together with a small paragraph about the requirements on duck arrays, but this should definitely be extended and moved somewhere more visible – maybe a `Extending xarray` document, together with the current section about custom accessors?

I also added a test, but it feels a bit strange to compare the result with the return value of a direct call to the object's `_repr_inline_`.

dcherian commented on doc/internals.rst:
```diff
@@ -42,6 +42,37 @@ xarray objects via the (readonly) :py:attr:`Dataset.variables
+In certain situations (e.g. when printing the variables of ``Dataset``), xarray
+will display the repr of a `duck array`_ in a single line, limiting it to a
+certain number of characters. If that would drop too much information, the
+`duck array`_ may define a ``_repr_inline_`` method:
```
```suggestion
`duck array`_ may define a ``_repr_inline_`` method that takes `max_width` (number of characters) as an argument:
```

keewis replied:
looks good, except that information is also part of the docstring of the example / skeleton. Should I remove the docstring and use this instead?

dcherian replied:
I think the redundancy is OK. I don't feel too strongly about this change so if you don't like it, feel free to ignore

max-sixty replied:
Agree it's good to have it in the prose
I think fine either way re repeating in docstring or not

keewis replied:
I'll keep both

dcherian commented on doc/whats-new.rst:
```diff
@@ -29,6 +29,9 @@ New Features
   :py:meth:`CFTimeIndex.__repr__` (:issue:`2416`, :pull:`4092`)
   `Aaron Spring <https://github.com/aaronspring>`_.
+- Use a wrapped array's ``_repr_inline_`` method to construct the collapsed ``repr``
+  of :py:class:`DataArray` and :py:class:`Dataset` objects. (:pull:`4248`).
```
could link to `internals.rst` here.

keewis commented:
should be ready for review again

max-sixty commented:
Perfect! Thanks for adding to the docs!

dcherian commented:
Thanks @keewis 

> get rid of our inline_dask_repr and inline_sparse_repr functions by pushing that upstream

I think this is a great idea!

shoyer commented on doc/internals.rst:
```diff
@@ -42,6 +42,38 @@ xarray objects via the (readonly) :py:attr:`Dataset.variables
+``shape``, ``dtype`` and ``ndim`` properties and the ``__array__``,
+``__array_ufunc__`` and ``__array_function__`` methods.
+
+In certain situations (e.g. when printing the collapsed preview of
+variables of a ``Dataset``), xarray will display the repr of a `duck array`_
+in a single line, truncating it to a certain number of characters. If that
+would drop too much information, the `duck array`_ may define a
+``_repr_inline_`` method that takes ``max_width`` (number of characters) as an
+argument:
```
It would be nice to add examples of what should and should not be included in `_repr_inline_`.

In particular, `shape` and `dtype` should *not* be included because xarray already adds those in a standard way.

For examples of what good `_inline_repr_` look like, we could show examples of the output for dask and sparse.
